homeworks/JohannesMerkt/1-halfedge/src/geometry/halfedge.ts

```diff
@@ -0,0 +1,248 @@
+// Created by Changkun Ou <https://changkun.de>.
+//
+// Use of this source code is governed by a GNU GPLv3 license that can be found
+// in the LICENSE file.
+
+import {Vertex, Edge, Face, Halfedge} from './primitive';
+import {Vector} from '../linalg/vec';
+import {Matrix} from '../linalg/mat';
+import {Quaternion} from '../linalg/quaternion';
+
+export class HalfedgeMesh {
+  // context is a transformation context (model matrix) that accumulates
+  // applied transformation matrices (multiplied from the left side) for
+  // the given mesh.
+  //
+  // context is a persistent status for the given mesh and can be reused
+  // for each of the rendering frames unless the mesh intentionally
+  // calls the resetContext() method.
+  context: Matrix;
+  color: Vector;
+  wireframe: Vector;
+
+  // The following four fields are the key fields to represent half-edge based
+  // meshes.
+  verts: Vertex[]; // a list of vertices
+  edges: Edge[]; // a list of edges
+  faces: Face[]; // a list of faces
+  halfedges: Halfedge[]; // a list of halfedges
+
+  /**
+   * constructor constructs the halfedge-based mesh representation.
+   *
+   * @param {string} data is a text string from an .obj file
+   */
+  constructor(data: string) {
+    // context is initialized as an identity matrix.
+    // prettier-ignore
+    this.context = new Matrix(
+      1, 0, 0, 0,
+      0, 1, 0, 0,
+      0, 0, 1, 0,
+      0, 0, 0, 1,
+    );
+    this.color = new Vector(0, 128, 255, 1);
+    this.wireframe = new Vector(125, 125, 125, 1);
+
+    // load .obj file
+    const indices: number[] = [];
+    const positions: Vector[] = [];
+    const lines = data.split('\n');
+    for (let line of lines) {
+      line = line.trim();
+      const tokens = line.split(' ');
+      switch (tokens[0].trim()) {
+        case 'v':
+          positions.push(
+            new Vector(
+              parseFloat(tokens[1]),
+              parseFloat(tokens[2]),
+              parseFloat(tokens[3]),
+              1
+            )
+          );
+          break;
+        case 'f':
+          // only load indices of vertices
+          for (let i = 1; i < tokens.length; i++) {
+            const vv = tokens[i].split('/');
+            indices.push(parseInt(vv[0]) - 1);
+          }
+          break;
+      }
+    }
+
+    this.verts = []; // an array of Vertex object
+    this.edges = []; // an array of Edge object
+    this.faces = []; // an array of Face object
+    this.halfedges = []; // an array of Halfedge object
+    this.buildMesh(indices, positions);
+  }
+
+  /**
+   * buildMesh builds half-edge based connectivity for the given vertex index buffer
+   * and vertex position buffer.
+   *
+   * @param indices is the vertex index buffer that contains all vertex indices.
+   * @param positions is the vertex buffer that contains all vertex positions.
+   */
+  buildMesh(indices: number[], positions: Vector[]) {
```

Comment:
Apart from preallocation as we have discussed this week, everything LGTM.

Reply:
Yes i noticed. In the second assignment my code was to slow so I had to rewrite the code just now.